### PR TITLE
upgrade: tokio 0.2 in deno_core_http_bench, take2

### PR DESCRIPTION
**Pull request by bartlomieju**
Trying to reland #3408

This time using async mutex

### CODE REVIEW

File: core/Cargo.toml [hunk 1]
```diff
@@ -28,4 +28,4 @@ path = "examples/http_bench.rs"
 
 # tokio is only used for deno_core_http_bench
 [dev_dependencies]
-tokio = "0.1.18"
+tokio = { version = "0.2.0", features = ["full"] }
```
Comment (ry) [inline core/Cargo.toml]: Isn’t it at 0.2.2 now?

Comment (bartlomieju) [reply core/Cargo.toml]: Oooops something got lost during rebase

File: core/libdeno.rs [hunk 2]
```diff
@@ -97,6 +97,8 @@ pub struct PinnedBufRaw {
 
 unsafe impl Send for PinnedBuf {}
 unsafe impl Send for PinnedBufRaw {}
+unsafe impl Sync for PinnedBuf {}
+unsafe impl Sync for PinnedBufRaw {}
```
Comment (ry) [inline core/libdeno.rs]: @piscisaureus is this ok?

File: core/examples/http_bench.rs [hunk 3]
```diff
@@ -273,17 +304,20 @@ fn op_read(
...
+
+  let fut = async move {
+    let nread = {
+      debug!("pre lock read");
+      let mut table = lock_resource_table().await;
+      debug!("post lock read");
+      let stream = table.get_mut::<TcpStream>(rid).ok_or_else(bad_resource)?;
+      let stream = &mut stream.0;
+      stream.read(&mut zero_copy_buf).await?
+    };
```
Comment (bartlomieju) [inline core/examples/http_bench.rs]: I'm not sure if this bit is correct. I believe that `MutexGuard` is held all the way until `stream.read(&mut zero_copy_buf).await?` completes which is not good. Probably should be implemented the same way as `Accept` above. Rewriting it as standalone future would render `unsafe impl Sync for PinnedBuf {}` not needed

File: core/examples/http_bench.rs [hunk 4]
```diff
@@ -294,17 +328,20 @@ fn op_write(
...
-  .and_then(move |nwritten| {
+
+  let fut = async move {
+    let nwritten = {
+      debug!("pre lock write");
+      let mut table = lock_resource_table().await;
+      debug!("post lock write");
+      let stream = table.get_mut::<TcpStream>(rid).ok_or_else(bad_resource)?;
+      let stream = &mut stream.0;
+      stream.write(&zero_copy_buf).await?
```
Comment (bartlomieju) [inline core/examples/http_bench.rs]: ditto

**Comment by bartlomieju** [comment]
@ry I run benchmarks locally using only 2 logical core (to replicate GitHub Actions) and was unable to replicate such performance hit (I get 85k req/s compared to ~76k req/s on single thread). Unfortunately I can't run `cargo instruments` anymore so I don't know if Isolate is being thrown between threads.

**Comment by bartlomieju** [comment]
@ry I merged latest `master` and reverted to `std::sync::Mutex`, so this PR effectively only updates Tokio in `deno_core_http_bench`. Do you think we can land it now?

**Pull request merged**